Implement postgres types for JSON, timestamps, and bytes

silasdavis opened the pull request:
This only has a basic unit test. I was going to add something to example package just so i could test it. is there anywhere else for an integration test?

I've given a nod to your suggestion that people may want different type aliases with my refactor of the type mapping. Currently this allows you to provide an alternative name and optional definition. I thought it might be useful to allow imports to external types to be dragged in but in some ways it's nice to keep it local.

fixes #45

adelsz commented:
Thanks @silasdavis, this looks really good!
I especially like the `TypeAlias` abstraction 👍 

There an no integration tests yet. It makes sense to add a JSON column/query to the `example` app, both as reference for users of json cols and as a basic smoke test. The generation unit tests should cover the rest for now.

Yes, it would be nice to allow users to provide a basic `pg-type -> type alias` map in the config. If an external type alias is needed, then the user can specify import statements that get added to all generated files. I suggest to think more about the config in a potential separate task/PR, to keep this PR focused on the internal type mapping.

adelsz commented on packages/cli/src/types.ts:
```diff
@@ -0,0 +1,126 @@
+  bytea: Bytes,
+});
+
+export type PostgresType = keyof typeof DefaultTypeMapping;
```
Lets keep the naming DB neutral to make porting this to other DBs a little easier.

adelsz commented on packages/cli/src/types.ts:
```diff
@@ -0,0 +1,126 @@
+import { keyof } from 'io-ts';
+
+export class TypeAlias {
+  constructor(
```
What do you think about adding an `inline: boolean` option here?
This will allow to inline core primitive type instead of reexporting them as aliases.
I think it makes sense to not alias `number`, `string` and `boolean` to reduce potential confusion for new users.

silasdavis replied:
I think I might be missing your meaning of inline. But if you omit the definition then no declaration is emitted so we don't provide some alias for 'boolean' etc. The idea is if you don't provide the optional declaration it works just like the old enum version did - then it just acts like a typed constant. I was thinking that a type name is the noop version of a TypeAlias.

adelsz replied:
Ah, right, disregard this. For some reason I thought all type aliases are reexported.

silasdavis commented:
After testing my draft changes I made a few changes:

1. My first approach duplicating type definitions for each query so I had to drag the type definitions up to the declaration file generation. This made it possible to implement imports which I wanted to do initially so I did.

2. Rejigged the `Type` types and renamed `TypeAccumulator` to `TypeAllocator`, which can now merge imports and type aliases and output a declaration header for the declaration file. 

3. I refactored some top-level generation functions into generator.ts. I did this initially because I thought it was a cleaner way to test them. I ended up not testing them directly. But I thought the refactor was cleaner (we read files in index.ts and do generation in generation.ts). this isn't strictly necessary though so I can reverse it if you like. Down the line I think we should test those functions though so generation.ts seems better to me.

4. I added a `allowUnmappedTypes: bool` option to `TypeAllocator`. I found this is was useful to me when working with unsupported postgres types (specifically enums). I could enable it in code and then fix up the generated files manually. Not sure if you want to expose this through in CLI config.

5. TypeAllocator supports imported types as well as aliased types - this could be used for custom types or types pgtyped doesn't currently generate. Again this could be added to CLI config possibly as a map of replacement types. 

I have found 4 and 5 have helped me to benefit from pgtyped on queries with unsupported types. So nice to have them available internally at least.

adelsz commented:
Thanks @silasdavis! All of the points make sense, 4 and 5 are definitely nice to have.

silasdavis commented on packages/cli/src/generator.ts:
```diff
@@ -176,6 +126,8 @@ export async function queryToTypeDeclarations(
     }
   }
 
+  await types.check();
```
This throws if there was an error (which currently leads to #53). Just thinking about the pattern for error handling you describe a preference for there. Perhaps this should just be logging the errors. Unmapped types already output `never`.

adelsz replied:
Lets remove this call for now and revisit error handling in #53

silasdavis commented on packages/example/sql/schema.sql:
```diff
@@ -1,49 +1,76 @@
-  registration_date DATE NOT NULL DEFAULT CURRENT_DATE
+CREATE TABLE users
+(
+    id                SERIAL PRIMARY KEY,
```
let me fix the autoformatting here

silasdavis commented on packages/example/docker-compose.yml:
```diff
@@ -17,5 +17,6 @@ services:
     build: .
     volumes:
       - ../../:/app
+    restart: always
```
@adelsz I have just noticed that for me this restarts these containers when docker restarts - apparently even after a restart. I could have sworn I've used this before without that effect. In any case I think `on-failure` would do the job here - the aim was just fro the example container to restart until postgres was ready. There are other ways also but I _thought_ this would be simpler

adelsz replied:
Agree, `on-failure` should be a better fit here. I should probably add a reconnect feature to the CLI at some point.
